### PR TITLE
[SPARK-4943][SQL] Allow table name having dot for db/catalog

### PR DESCRIPTION
**Pull request by alexliu68**
The pull only fixes the parsing error and changes API to use tableIdentifier. Joining different catalog datasource related change is not done in this pull.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala [hunk 1]
```diff
@@ -178,10 +178,23 @@ class SqlParser extends AbstractSparkSQLParser {
...
+      | ident ~ ("." ~> ident) ~ ("." ~> ident) ~ (opt(AS) ~> opt(ident)) ^^ {
+        case reserveName1 ~ dbName ~ tableName ~ alias =>
+          UnresolvedRelation(IndexedSeq(tableName, dbName, reserveName1), alias)
+      }
+      | ident ~ ("." ~> ident) ~ (opt(AS) ~> opt(ident)) ^^ {
+          case dbName ~ tableName ~ alias =>
+            UnresolvedRelation(IndexedSeq(tableName, dbName), alias)
+        }
+      | ident ~ (opt(AS) ~> opt(ident)) ^^ {
+          case tableName ~ alias => UnresolvedRelation(IndexedSeq(tableName), alias)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala]: I think this should probably just be a single rule.  With something like: `repsep(ident, ",")` for the table identifier.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 2]
```diff
@@ -115,43 +101,41 @@ class SimpleCatalog(val caseSensitive: Boolean) extends Catalog {
 trait OverrideCatalog extends Catalog {
 
   // TODO: This doesn't work when the database changes...
-  val overrides = new mutable.HashMap[(Option[String],String), LogicalPlan]()
+  val overrides = new mutable.HashMap[String, LogicalPlan]()
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: Why collapse this to a `String`?  What about a table with a name like the following:
```
SELECT * FROM `database`.`table.name`
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 3]
```diff
@@ -28,77 +28,63 @@ trait Catalog {
...
-  def unregisterTable(databaseName: Option[String], tableName: String): Unit
+  def unregisterTable(tableIdentifier: Seq[String]): Unit
 
   def unregisterAllTables(): Unit
 
-  protected def processDatabaseAndTableName(
-      databaseName: Option[String],
-      tableName: String): (Option[String], String) = {
+  protected def processTableIdentifier(tableIdentifier: Seq[String]):
+      Seq[String] = {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: I don't think this needs to be wrapped, does it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 4]
```diff
@@ -28,77 +28,63 @@ trait Catalog {
...
 
-  def tableExists(db: Option[String], tableName: String): Boolean
+  def tableExists(tableIdentifier: Seq[String]): Boolean
 
   def lookupRelation(
-    databaseName: Option[String],
-    tableName: String,
-    alias: Option[String] = None): LogicalPlan
+                      tableIdentifier: Seq[String],
+                      alias: Option[String] = None): LogicalPlan
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: Indent wrapped arguments 4 spaces.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala [hunk 5]
```diff
@@ -28,77 +28,63 @@ trait Catalog {
...
-      tableName: String,
+      tableIdentifier: Seq[String],
       alias: Option[String] = None): LogicalPlan = {
-    val (dbName, tblName) = processDatabaseAndTableName(databaseName, tableName)
-    val table = tables.getOrElse(tblName, sys.error(s"Table Not Found: $tableName"))
-    val tableWithQualifiers = Subquery(tblName, table)
+    val tableIdent = processTableIdentifier(tableIdentifier)
+    val tableFullName = tableIdent.mkString(".")
+    val table = tables.getOrElse(tableFullName, sys.error(s"Table Not Found: $tableFullName"))
+    val tableWithQualifiers = Subquery(tableIdent.head, table)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: `tableIdent.head`?  Should it be `last`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala [hunk 6]
```diff
@@ -386,6 +386,13 @@ private[hive] object HiveQl {
     (db, tableName)
   }
 
+  protected def extractTableIdent(tableNameParts: Node): IndexedSeq[String] = {
+    tableNameParts.getChildren.map { case Token(part, Nil) => cleanIdentifier(part) } match {
+      case Seq(tableOnly) => IndexedSeq(tableOnly)
+      case Seq(databaseName, table) => IndexedSeq(table, databaseName)
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala]: I'd add a default case here that does something like `case other => sys.error(s"Hive only supports tables names like 'tableName' or 'databaseName.tableName', found '$other'")`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 7]
```diff
@@ -251,6 +257,26 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
...
+      databaseName: Option[String],
+      tableName: String): (Option[String], String) = {
+    if (!caseSensitive) {
+      (databaseName.map(_.toLowerCase), tableName.toLowerCase)
+    } else {
+      (databaseName, tableName)
+    }
+  }
+
+  protected def processDatabaseAndTableName(
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Are either of these functions used anymore?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 8]
```diff
@@ -57,18 +58,23 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
...
 
   def lookupRelation(
-      db: Option[String],
-      tableName: String,
+      tableIdentifier: Seq[String],
       alias: Option[String]): LogicalPlan = synchronized {
+    val tableIdent = processTableIdentifier(tableIdentifier)
     val (databaseName, tblName) =
-      processDatabaseAndTableName(db.getOrElse(hive.sessionState.getCurrentDatabase), tableName)
+      (tableIdent.lift(1).getOrElse(hive.sessionState.getCurrentDatabase), tableIdent.head)
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: I did not know that you could `lift` a `Seq` :)

**Comment by marmbrus** [comment]
Thanks for doing this cleanup!  Overall I think this looks pretty good.  I made a few small comments.  One other very minor nit.  It seems a little odd to me to be using `IndexedSeq` everywhere instead of just `Seq` or `:: Nil`.

**Comment by marmbrus** [comment]
Also, can you fix the PR title to avoid truncating.

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 3 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: agreed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 4 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: agreed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 5 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: I store the seq in reversed order, so table name is at head.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala (hunk 6 above)
Comment (alexliu68) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala]: agreed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 5 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: Oh, sorry that was not obvious to me, and seems pretty confusing.  Why not store it in the order the user gave it to you?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 7 above)
Comment (alexliu68) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: It's only used by hive table creation

**Comment by alexliu68** [comment]
I use IndexedSeq instead Seq for IndexedSeq is faster to access element by lift. I can revert them back to Seq.

**Comment by marmbrus** [comment]
Its minor, but it seems unlikely that it it noticeably faster in this instance.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 5 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: By reversed order, we know database name will be at 2nd, so it's easy to access it by using lift(1)

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: That does not seem worth the confusion of having it in reverse order.  Also looking at that code more closely, we would silently discard extra part of the table identifier right? Seems like we should explicitly handle the cases where there are 1, 2, and more elements of the table identifier.

Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: agreed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala (hunk 1 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SqlParser.scala]: I change it to rep1sep(ident, ".")

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala (hunk 2 above)
Comment (alexliu68) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala]: restore it to (Option[String],String)

**Comment by yhuai** [comment]
@alexliu68 Is there any particular reason that `org.apache.spark.sql.hive.HiveMetastoreCatalog#createTable` does not take a `tableIdentifier: Seq[String]`?

**Comment by alexliu68** [comment]
I only change selection to use tableIdentifier for joining tables. I keep createTable no change to minimize API changes. createTable could also use tableIdentifier to create catalog/cluster/database level table, but I leave it for future development if we need it.

**Pull request closed without merging**